synapse/federation/federation_server.py

```diff
@@ -169,6 +170,33 @@ async def on_incoming_transaction(
 
         logger.debug("[%s] Got transaction", transaction_id)
 
+        # Reject malformed transactions early: reject if too many PDUs/EDUs
+        if len(transaction.pdus) > 50 or (  # type: ignore
+            hasattr(transaction, "edus") and len(transaction.edus) > 100  # type: ignore
+        ):
+            logger.info("Transaction PDU or EDU count too large. Returning 400")
+            return 400, {}
```

Comment:
I moved this up here from `_handle_incoming_transaction` to keep a sytest (which was sending concurrent transactions) happy, but it seems like generally a good thing to do.

Reply:
I agree that it seems best to do this as early as possible. 👍